tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/api/RecoveryResource.java

```diff
@@ -0,0 +1,188 @@
+        URI lra;
+        Response response;
+        
+        response = lraTestService.getTCKSuiteTarget().path(RecoveryResource.RECOVERY_RESOURCE_PATH)
+                .path(timeout ? RecoveryResource.REQUIRED_TIMEOUT_PATH : RecoveryResource.REQUIRED_PATH)
+                .request().put(Entity.text(""));
+
+        lra = URI.create(response.readEntity(String.class));
+
+        Response.ResponseBuilder responseBuilder = Response.ok(lra);
+        if (timeout) {
+            responseBuilder.header(TIMEOUT_HEADER, lraTckConfigBean.getShortConsistencyDelay());
+        }
+        
+        return responseBuilder.build();
+    }
+
+    /**
+     * Cancels the supplied LRA and verifies that all required actions have been performed
+     * 
+     * @param lraId lra id of the LRA to be cancelled
+     * @return a {@link Response} object containing the optional error message 
+     */
+    @GET
+    @Path(PHASE_2)
+    public Response phase2(@HeaderParam(LRA_HEADER) URI lraId) {
+        lraTestService.getLRAClient().cancelLRA(lraId);
+        lraTestService.applyShortConsistencyDelay();
+
+        // assert compensate has been called
+        int compensations = lraMetricService.getMetric(LRAMetricType.Compensated, lraId, RecoveryResource.class.getName());
+        if (compensations != 1) {
+            return assertFailedResponse("Compensate on restarted service should have been called. Was " + compensations);
+        }
+
+        // assert after LRA has been called
+        int afterLRACalls = lraMetricService.getMetric(LRAMetricType.Cancelled, lraId, RecoveryResource.class.getName());
+        if (afterLRACalls != 1) {
+            return assertFailedResponse("After LRA with Cancelled status should have been called. Was " + afterLRACalls);
+        }
+
+        return Response.ok().build();
+    }
+
+    private Response assertFailedResponse(String message) {
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(message).build();
+    }
+
+    @GET
+    @Path(TRIGGER_RECOVERY)
+    public Response triggerRecovery(@HeaderParam(LRA_HEADER) URI lraId) {
+        lraRecoveryService.triggerRecovery(lraId);
+        return Response.ok().build();
+    } 
+
+    // LRA participant methods
+
+    @PUT
+    @Path(REQUIRED_PATH)
+    @LRA(value = LRA.Type.REQUIRED, end = false)
+    public Response requiredLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.ok(lraId).build();
+    }
+
+    @PUT
+    @Path(REQUIRED_TIMEOUT_PATH)
+    @LRA(value = LRA.Type.REQUIRED, end = false, timeLimit = 500, timeUnit = ChronoUnit.MILLIS)
+    public Response requiredTimeoutLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.ok(lraId).build();
+    }
+
+    @PUT
+    @Path("/compensate")
+    @Compensate
+    public Response compensate(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        LOGGER.info("RecoveryResource compensating " + lraId);
```

Comment:
Another "just a comment": adding logging at info for standard execution logic makes logging output not useful. At most (if at all) it should be at trace level.